src/AppInstallerCommonCore/Fonts.cpp

```diff
@@ -0,0 +1,177 @@
+
+                if (const auto localLoader = loader.try_query<IDWriteLocalFontFileLoader>()) {
+                    UINT32 pathLength;
+                    THROW_IF_FAILED(localLoader->GetFilePathLengthFromKey(fontFileReferenceKey, fontFileReferenceKeySize, &pathLength));
+                    pathLength += 1; // Account for the trailing null terminator during allocation.
+
+                    wchar_t* path = new wchar_t[pathLength];
+                    THROW_IF_FAILED(localLoader->GetFilePathFromKey(fontFileReferenceKey, fontFileReferenceKeySize, &path[0], pathLength));
+                    filePaths.push_back(std::move(std::wstring(path)));
+                }
+            }
+
+            return filePaths;
+        }
+
+        std::wstring GetLocalizedStringFromFont(const wil::com_ptr<IDWriteLocalizedStrings>& localizedStringCollection)
+        {
+            std::vector<std::string> locales = AppInstaller::Locale::GetUserPreferredLanguages();
+            std::wstring preferredLocale = Utility::ConvertToUTF16(!locales.empty() ? locales[0] : "en-US");
+
+            UINT32 index;
+            BOOL exists;
+            // TODO: Aggregate available locales and find best alternative locale if preferred locale does not exist.
+            if (FAILED(localizedStringCollection->FindLocaleName(preferredLocale.c_str(), &index, &exists)) || !exists)
+            {
+                index = 0;
+            }
+
+            UINT32 length = 0;
+            THROW_IF_FAILED(localizedStringCollection->GetStringLength(index, &length));
+            length += 1; // Account for the trailing null terminator during allocation.
+
+            wchar_t* localizedString = new wchar_t[length];
+            THROW_IF_FAILED(localizedStringCollection->GetString(index, localizedString, length));
+            return std::wstring(localizedString);
+        }
+
+        std::wstring GetFontFaceName(const wil::com_ptr<IDWriteFont>& font)
+        {
+            wil::com_ptr<IDWriteLocalizedStrings> faceNames;
+            THROW_IF_FAILED(font->GetFaceNames(faceNames.addressof()));
+            return GetLocalizedStringFromFont(faceNames);
+        }
+
+        std::wstring GetFontFamilyName(const wil::com_ptr<IDWriteFontFamily>& fontFamily)
+        {
+            wil::com_ptr<IDWriteLocalizedStrings> familyNames;
+            THROW_IF_FAILED(fontFamily->GetFamilyNames(familyNames.addressof()));
+            return GetLocalizedStringFromFont(familyNames);
+        }
+
+        std::wstring GetFontFaceVersion(const wil::com_ptr<IDWriteFont>& font)
+        {
+            wil::com_ptr<IDWriteLocalizedStrings> fontVersion;
+            BOOL exists;
+            THROW_IF_FAILED(font->GetInformationalStrings(DWRITE_INFORMATIONAL_STRING_VERSION_STRINGS, fontVersion.addressof(), &exists));
+            if (!exists)
+            {
+                return {};
+            }
+
+            std::string value = Utility::ConvertToUTF8(GetLocalizedStringFromFont(fontVersion));
```

Comment:
Why are you converting to UTF8 if you are going to convert back to UTF16 to return it?

Reply:
Maybe a better question is just: Why are you converting back to UTF16 to return a value?  Why not our `Version` object?

https://learn.microsoft.com/en-us/typography/opentype/spec/name

> Version string. Should begin with the pattern “Version <number>.<number>” (upper case, lower case, or mixed, with a space between “Version” and the number).
>
> The string must contain a version number of the following form: one or more digits (0-9) of value less than 65,535, followed by a period, followed by one or more digits of value less than 65,535. Any character other than a digit will terminate the minor number. A character such as “;” is helpful to separate different pieces of version information.
> 
> The first such match in the string can be used by installation software to compare font versions. Some installers could require the string to start with “Version ”, followed by a version number as above.

**Should** start with "Version ".
**Must** contain a major.minor version component.

It sounds like the actual algorithm prescribed is (technically anywhere that it fails means start trying again, but we don't have to be so forgiving):
1. Find first digit
2. Read digits until non-digit
3. If anything except a `.`, error
4. If value >= 65535, error
5. Skip the `.`
6. Read digits until non-digit
7. If value >= 65535, error

But I think we can simplify it a bit:
1. Split on `.`
2. Ensure at least 2 parts
3. Part 0: Find first digit, parse number, ensure remainder consumed
4. Part 1: Ensure starts with digit, parse number, toss remainder
5. `Version` is the two parts with no "other"

I would agree that throwing out the preceding part (which is most likely "Version ") makes sense.  We could keep everything after the minor part as the "other" if we wanted to, but since this prescribes it to have no meaning, we probably shouldn't.

I would create an `OpenTypeFontVersion` (derived from `Version` of course) that encoded this, allowing a consistent construction.  This also makes it easy to test various strings, as this seems to be completely open to whatever arbitrary input someone wants to put in.

Reply:
With the new preamble stripping, a derived type may not even be necessary